src/lib.rs

```diff
@@ -186,67 +188,72 @@ impl Progress {
     /// **Note 2:** If your program is not being run in a terminal, an initial
     /// empty bar will be printed but never refreshed.
     pub fn draw(&mut self, bar: &Bar) {
+        self.draw_impl(bar, false);
+
+        // Very important, or the output won't appear fluid.
+        let _ = self.out.flush();
+    }
+
+    /// Actually draw a particular [`Bar`].
+    ///
+    /// When `force` is true draw the bar at the current cursor position and
+    /// advance the cursor one line.
+    ///
+    /// This function does not flush the output stream.
+    fn draw_impl(&mut self, bar: &Bar, force: bool) {
         // If there is no legal width value present, that means we aren't
         // running in a terminal, and no rerendering can be done.
         if let Some((term_width, term_height)) = self.size {
             let pos = self.bars.len() - bar.0;
+            let mut b = &mut self.bars[bar.0];
+            let cur_percent = (100 * b.curr as u64) / (b.total as u64);
+            // For a newly cancelled bar `diff` is equal to 100.
+            let diff = cur_percent - b.prev_percent as u64;
 
             // For now, if the progress for a particular bar is slow and drifts
             // past the top of the terminal, redrawing is paused.
-            if pos < term_height {
-                let mut b = &mut self.bars[bar.0];
+            if (pos < term_height && diff >= 1) || force {
```

Comment:
If `force` still necessary?

Reply:
Yes, we need it because we depend on the position relative to the end of the user text because we don't know how many lines they wrote.

And I get that providing a good quality/looking scrollback is not a goal of this library.
As a user, for me, it's still important that it's at least lossless, so I can scroll up and look through the bars whether `test.png` was downloaded or only `foo.jpg` and `bar.gif`. And this use of `force` additionally ensures that a bar is never overwritten.